providers/websockets/binance/messages.go

```diff
@@ -0,0 +1,255 @@
+//			"x": "0.0009",      // First trade(F)-1 price (first trade before the 24hr rolling window)
+//			"c": "0.0025",      // Last price
+//			"Q": "10",          // Last quantity
+//			"b": "0.0024",      // Best bid price
+//			"B": "10",          // Best bid quantity
+//			"a": "0.0026",      // Best ask price
+//			"A": "100",         // Best ask quantity
+//			"o": "0.0010",      // Open price
+//			"h": "0.0025",      // High price
+//			"l": "0.0010",      // Low price
+//			"v": "10000",       // Total traded base asset volume
+//			"q": "18",          // Total traded quote asset volume
+//			"O": 0,             // Statistics open time
+//			"C": 86400000,      // Statistics close time
+//			"F": 0,             // First trade ID
+//			"L": 18150,         // Last trade Id
+//			"n": 18151          // Total number of trades
+//	}
+//
+// ref: https://developers.binance.com/docs/binance-spot-api-docs/web-socket-streams#individual-symbol-ticker-streams
+type TickerMessageResponse struct {
+	Data struct {
+		// Ticker is the symbol.
+		Ticker string `json:"s"`
+		// LastPrice is the last price.
+		LastPrice string `json:"c"`
+		// StatisticsCloseTime is the statistics close time.
+		//
+		// Note: This is unused but is included since json.Unmarshal requires all fields with same character but different casing
+		// to be present.
+		StatisticsCloseTime int64 `json:"C"`
+	} `json:"data"`
+}
+
+// NewSubscribeRequestMessage returns a set of messages to subscribe to the Binance websocket. This will
+// subscribe each instrument to the aggregate trade and ticker streams.
+func (h *WebSocketHandler) NewSubscribeRequestMessage(instruments []string) ([]handlers.WebsocketEncodedMessage, error) {
+	if len(instruments) == 0 {
+		return nil, fmt.Errorf("no instruments to subscribe to")
+	}
+
+	msgs := make([]handlers.WebsocketEncodedMessage, 0)
+	for _, instrument := range instruments {
+		// Generate a random ID.
+		id := h.GenerateID()
+		msg, err := json.Marshal(SubscribeMessageRequest{
+			Method: string(SubscribeMethod),
+			Params: []string{
+				fmt.Sprintf("%s%s%s", strings.ToLower(instrument), Separator, string(AggregateTradeStream)),
+				fmt.Sprintf("%s%s%s", strings.ToLower(instrument), Separator, string(TickerStream)),
+			},
+			ID: id,
+		})
+		if err != nil {
+			return nil, fmt.Errorf("failed to marshal subscribe message: %w", err)
+		}
+
+		// Set the IDs
+		h.SetIDForInstruments(id, []string{instrument})
+		msgs = append(msgs, msg)
+	}
+
+	return msgs, nil
+}
+
+// SetIDForInstruments sets the ID for the given instruments. This is used to set the ID for the
+// instruments that are being subscribed to.
+func (h *WebSocketHandler) SetIDForInstruments(id int64, instruments []string) {
+	h.messageIDs[id] = instruments
+}
+
+// GenerateID generates a random ID for the message.
+func (h *WebSocketHandler) GenerateID() int64 {
+	for {
+		id := rand.Int63() + 1
+		if _, ok := h.messageIDs[id]; !ok {
+			return id
+		}
+	}
```

Comment:
alternatively you could store an int64 on the handler, and bump it everytime you need a new ID? might be a bit simpler, but also might be optimization mogging you with this comment lol

Reply:
should have done that to begin with, this for loop scares me anyways lol